File: src/org/cerion/symcalc/expression/function/integer/Binomial.java
package org.cerion.symcalc.expression.function.integer;

import org.cerion.symcalc.expression.Expr;
import org.cerion.symcalc.expression.FunctionExpr;
import org.cerion.symcalc.expression.number.IntegerNum;

public class Binomial extends FunctionExpr {

	public Binomial(Expr ...e) {
		super(FunctionType.BINOMIAL, e);
	}
	
	@Override
	protected Expr evaluate() {
		
		if(get(0).isInteger() && get(1).isInteger())
		{
			IntegerNum n = (IntegerNum)get(0);
			IntegerNum k = (IntegerNum)get(1);
			
			// n! / k!(n-k)!
			IntegerNum n1 = n.factorial();
			IntegerNum n2 = n.subtract(k);
			n2 = n2.factorial();
			IntegerNum n3 = k.factorial();
			n2 = n2.multiply(n3);
			
			return n1.divide(n2);
		}
 			
		return this;
	}
	
}

File: src/org/cerion/symcalc/expression/function/integer/Binomial.kt
package org.cerion.symcalc.expression.function.integer

import org.cerion.symcalc.exception.ValidationException
import org.cerion.symcalc.expression.Expr
import org.cerion.symcalc.expression.FunctionExpr
import org.cerion.symcalc.expression.number.NumberType

class Binomial(vararg e: Expr) : FunctionExpr(FunctionExpr.FunctionType.BINOMIAL, *e) {

    override fun evaluate(): Expr {
        val n = get(0).asInteger()
        val k = get(1).asInteger()

        // n! / k!(n-k)!
        val n1 = Factorial(n).eval().asInteger()
        var n2 = n - k
        n2 = Factorial(n2).eval().asInteger()
        val n3 = Factorial(k).eval().asInteger()
        n2 = n2.times(n3)

        return n1 / n2
    }

    @Throws(ValidationException::class)
    override fun validate() {
        validateParameterCount(2)
        validateNumberType(0, NumberType.INTEGER)
        validateNumberType(1, NumberType.INTEGER)
    }
}